feat: add MCP tool filtering support

devtalker opened the pull request:
## Add MCP Tool Filtering Support

This PR implements tool filtering capabilities for MCP servers, addressing multiple community requests for this feature.

### Problem
Currently, Agent SDK automatically fetches all available tools from MCP servers without the ability to select specific tools. This creates several issues:
- Unwanted tools occupy LLM context and affect tool selection accuracy
- Security concerns when limiting tool access scope
- Tool name conflicts between multiple servers
- Need for different tool subsets across different agents

### Solution
Implements a two-level filtering system:

**Server-level filtering:**
```python
server = MCPServerStdio(
    params={"command": "npx", "args": ["-y", "@modelcontextprotocol/server-filesystem", "/path"]},
    allowed_tools=["read_file", "write_file"],  # whitelist
    excluded_tools=["delete_file"]  # blacklist
)
```

**Agent-level filtering:**
```python
agent = Agent(
    name="Assistant",
    mcp_servers=[server1, server2],
    mcp_config={
        "allowed_tools": {"server1": ["read_file", "write_file"]},
        "excluded_tools": {"server2": ["dangerous_tool"]}
    }
)
```

### Features
- ✅ Server-level `allowed_tools`/`excluded_tools` parameters for all MCP server types
- ✅ Agent-level filtering via `mcp_config` 
- ✅ Hierarchical filtering (server-level first, then agent-level)
- ✅ Comprehensive test coverage (8 test cases)
- ✅ Updated documentation with examples

### Related Issues
#376, #851, #830, #863

### Testing
All existing tests pass + 8 new test cases covering various filtering scenarios.

rm-openai reviewed (changes requested):
I like this. Would however prefer an interface that looked something like this:
```
ToolFilterCallable = Callable[[ToolFilterContext, Tool], MaybeAwaitable[bool]]
class ToolFilterStatic(TypedDict):
  allowed_tool_names: NotRequired[list[str]]
  blocked_tool_names: NotRequired[list[str]]
```
and then the MCPServer class would take `tool_filter: ToolFilterCallable | ToolFilterStatic | None`

thoughts?

devtalker commented:
@rm-openai Thanks for your suggestions!
I like this approach! It's very similar to the is_enabled pattern we use in function tools, which makes the API more consistent and intuitive. Supporting a callable for dynamic filtering gives us much more flexibility - we can make runtime decisions based on various conditions, just like we do with function tools. I'll update the PR to implement this design. 🙏

devtalker commented:
Hi [@rm-openai](https://github.com/rm-openai)! 👋

Thank you for the detailed feedback! I've completely redesigned the implementation following your suggested interface pattern. 

## 🔄 Major Changes in This Update:

**1. Implemented Your Suggested Interface:**
- Added `ToolFilterCallable = Callable[[ToolFilterContext, Tool], MaybeAwaitable[bool]]`
- Created `ToolFilterStatic` TypedDict with `allowed_tool_names` and `blocked_tool_names`
- MCPServer classes now accept `tool_filter: ToolFilterCallable | ToolFilterStatic | None`

**2. Enhanced Dynamic Filtering Support:**
The new interface supports powerful dynamic filtering capabilities:

```python
# Context-aware filtering
def context_filter(context: ToolFilterContext, tool) -> bool:
    # Access agent information
    agent_name = context.agent.name
    # Access server information  
    server_name = context.server_name
    # Custom logic based on runtime context
    return some_dynamic_logic(agent_name, server_name, tool)

# Async filtering for complex operations
async def async_filter(context: ToolFilterContext, tool) -> bool:
    result = await some_async_check(context, tool)
    return result

server = MCPServerStdio(
    params={"command": "npx", "args": [...]},
    tool_filter=context_filter  # or async_filter
)
```

**3. Simplified Static Filtering:**
```python
from agents.mcp import create_static_tool_filter

server = MCPServerStdio(
    tool_filter=create_static_tool_filter(
        allowed_tool_names=["read_file", "write_file"],
        blocked_tool_names=["delete_file"]
    )
)
```

**4. Architecture Improvements:**
- Removed Agent-level filtering complexity as discussed
- Moved all filtering logic to the MCP server layer for better separation of concerns
- Maintains full backward compatibility

Would you be able to take another look when you have a moment? Thanks again for the great guidance! 🙏

rm-openai reviewed (changes requested):
sorry for late review. Generally lgtm, just remove the nullability plz

devtalker commented on src/agents/agent.py:
```diff
@@ -256,14 +256,18 @@ async def get_prompt(
 
-    async def get_mcp_tools(self) -> list[Tool]:
+    async def get_mcp_tools(
+        self, run_context: RunContextWrapper[TContext] | None = None
```
okey, the inconsistency is weird. originally kept it nullable for direct `agent.get_mcp_tools()` calls but API consistency is more important. will make it required like `get_all_tools`.

devtalker replied:
users might have existing calls like `agent.get_mcp_tools()` without parameters, so making `run_context` required would be a breaking change for them. This needs to be weighed carefully.

devtalker commented:
> sorry for late review. Generally lgtm, just remove the nullability plz

Thanks for the review!  I've removed the nullability as suggested - both `run_context` and `agent` are now required parameters for API consistency.

Just wanted to double-check - are you comfortable with the breaking change this introduces? 🤔 You have more experience with this, so I'll follow your guidance.

All tests pass and ready for re-review! ✅

devtalker commented:
@seratch hi, I've noticed that the automated tests are failing, but I'm not sure about the specific reason. Could you please advise on how to fix this issue?  🤔 I've observed that this syntax is also being used in other modules of the codebase.
